design/ratelimit-design-new.md

```diff
@@ -0,0 +1,397 @@
+
+For local rate limiting, the rate limit itself is defined inline in the `RateLimitPolicy`, as token bucket properties.
+Since local rate limits apply *per Envoy*, each pod in the Envoy daemon set will have a token bucket with these properties.
+
+For global rate limiting, per Envoy's API, the `RateLimitPolicy` only defines a list of descriptors to send to an external RLS for a given request.
+All [descriptor entries defined by Envoy](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#config-route-v3-ratelimit-action) will be supported except `metadata` and `dynamic_metadata`.
+
+### Contour Configuration
+If using global rate limiting, the details of the external RLS must be configured with Contour.
+First, an `ExtensionService` must be defined for the RLS with the cluster-level details for the RLS itself.
+For example:
+```yaml
+apiVersion: projectcontour.io/v1alpha1
+kind: ExtensionService
+metadata:
+  namespace: projectcontour
+  name: ratelimit
+spec:
+  protocol: h2
+  services:
+    - name: ratelimit
+      port: 8081
+  timeoutPolicy:
+    # sets  the "timeout" property on the Envoy rate limit service
+    # config.
+    response: 50ms
+```
+
+Second, a new block will be added to the Contour config file to identify the RLS extension service and provide some additional parameters:
+```yaml
+rateLimitService:
+  # extensionService contains the namespace/name of the RLS extension service.
+  extensionService: projectcontour/ratelimit
+  # domain is passed to the RLS for all rate limit requests.
+  # Defaults to "contour".
+  domain: contour
+  # failOpen defines whether to fail open or closed. If false, then if the RLS
+  # cannot be reached or does not return a valid rate limit decision within the
+  # specified timeout, the client will receive a 429 response for their request.
+  #
+  # This sets the "denyOnFailure" field in the Envoy config.
+  failOpen: true
+```
+
+### Rate Limit Flows
+
+- First, Envoy applies local rate limiting to incoming requests.
+- If there are no tokens left in the relevant token bucket(s), a `429 Too Many Requests` response is returned.
+- If there was no local rate limiting defined, or there is a token available, Envoy proceeds to global rate limiting.
+- 1+ descriptors, each of which is an ordered list of key/value pairs, are generated for the request based on the global rate limit configuration.
+- The descriptors are sent to the external RLS via gRPC.
+- The external RLS makes a rate limit decision for the request, based on the descriptors it receives.
+- The external RLS returns a response indicating whether the client request should be rate-limited or not.
+- If the request should be rate limited, a `429 Too Many Requests` response is returned to the client along with the `x-envoy-ratelimited` header.
+- If the request should not be rate limited, it is routed to the appropriate upstream cluster and proceeds as normal.
+
```

Comment:
I'd like to see some thoughts about what and how we communicate to the different personas, both cluster admin and application owner.

In the case of the application owner, it would be nice to be able to tell if I've hit the rate limit, or at least that the rate limit is configured and working correctly for my service.

In the case of the cluster operator, it seems like it's useful to be able to find the services that are consuming the rate limit, ideally with some information about the passed descriptors.

Basically, where is the status surfaced, and what can we make available? I suspect that getting information out of the RLS is not possible currently, but at least specifying what a HTTPProxy user should expect to see if there is a ratelimit seems useful.

In general, I'd like to see a little more info about failure cases. If the RLS fails, how do we know? Does Contour know at all?

It's okay to put those as open questions if we have to, but I feel like a extension design isn't complete without some detail about the UX of consuming it.

We probably will need to see if there's some metrics you can pull from Envoy about this, and include some information about that in the guide page as well.

Reply:
> In this case, it's these two fields here: domain and failOpen, and when we do tracing (ie #399 ) there will be other service-specific things, that will be the same for every reference of that service.

Without checking how Envoy configures this, I'd suggest that these are properties of the vhost, not properties of the rate limiting service. Just like auth, there is probably some transition where vhosts move from their own rate limiting to the platform one and would twiddle these kinds of knobs.

Reply:
Fair point, I should probably have checked. I agree that for things that are per-vhost, the config should live there, but for things that are properties of the service itself, they should end up on the ExtensionService object.

Reply:
> In general, I'd like to see a little more info about failure cases. If the RLS fails, how do we know? Does Contour know at all?

I'm still confused about how we push status info into the ExtensionService when the external service is a black box (this applies to auth as well). Is the operator implementing the external service supposed to write status back to the ExtensionService? 

I see two stats for RateLimiting that might be useful to understand if the rate limit was hit, otherwise you might be able to scrape the metrics for specific HTTP status codes to understand? 
```
upstream_rq_retry_backoff_exponential | Counter | Total retries using the exponential backoff strategy
upstream_rq_retry_backoff_ratelimited | Counter | Total retries using the ratelimited backoff strategy
```

Reply:
> > In this case, it's these two fields here: domain and failOpen, and when we do tracing (ie #399 ) there will be other service-specific things, that will be the same for every reference of that service.
> 
> Without checking how Envoy configures this, I'd suggest that these are properties of the vhost, not properties of the rate limiting service. Just like auth, there is probably some transition where vhosts move from their own rate limiting to the platform one and would twiddle these kinds of knobs.

The challenge here is that, if we want to support rate limiting for non-TLS vhosts, then the filter can't be per-vhost, since all non-TLS vhosts share the same HTTP connection mananager. That's why it's not at the vhost level.

If we decide not to support non-TLS vhosts, or can come up with some alternate way of getting around the single HCM/filter constraint, then these params could move under the vhost.

I'm open to ideas here!

Reply:
As far as status, metrics are emitted by Envoy with rate limit info:

(Global) https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/rate_limit_filter#statistics
(Local) https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/local_rate_limit_filter#statistics

For global, looking at those should give a picture of if the service is configured correctly, as well as if requests are being limited. Do we want Contour to be scraping these and using them to populate info in the API somewhere?

Reply:
Added some details on the stats that Envoy exposes. Given that we don't scrape Envoy metrics to pull into Contour anywhere else, I'm not going to propose doing that now; the user can pull those into a Grafana dashobard from Envoy directly to observe the runtime behavior. We can pursue this separately if we want, though we'd probably want to start with more fundamental metrics around traffic serving, rather than doing it just for rate limiting, IMHO.